File: sync/lib/sync/backend/mailbox.ex
defmodule Sync.Backend.Mailbox do
  @moduledoc """
  Write-once slots for device pairing, and nothing more.

  Two devices that want to join the same vault may have no way to reach each
  other — a browser tab cannot open a UDP socket, so it can never be an iroh
  peer. The relay already stands between them, so it can carry the handshake.
  This is the smallest shape that lets it, while keeping the relay exactly as
  ignorant as it is for everything else it stores.

  ## Layout

      <root>/<rendezvous>/<session>/<slot>

  A **rendezvous** is named by 32 unguessable bytes the host mints. A **session**
  is named by 32 bytes a joiner mints; a rendezvous holds several because a
  joiner that spoils one (or an attacker that squats one) must be able to start
  a clean one. A session holds at most the six named slots of the handshake.

  ## What the relay knows

  Nothing. Every slot body is either a SPAKE2 message — which by construction
  reveals nothing testable about the pairing code — or a ciphertext sealed under
  a key derived from that code. The rendezvous id is not derived from the bucket
  id and does not appear anywhere else, so the relay cannot tell which of the
  buckets it stores a pairing belongs to, or whether it stores it at all.

  ## Write-once, and why the relay has to enforce it

  A slot that already holds bytes is never overwritten; the second writer gets
  `:exists` and the stored body is untouched. That is not about tamper-detection
  — the handshake's own confirmations are MACs over the full transcript, so
  rewriting a message fails whatever the relay does. It is about *squatting*:
  anyone who knows the rendezvous can take the slot the host was about to write,
  and a host that shrugged and carried on would verify a confirmation against a
  transcript it never wrote, spending one of its three attempts. Three of those
  retire a pairing code without the squatter guessing a digit. Refusing the write
  is what makes the host able to notice and bail out for free.

  ## Separate root, deliberately

  Mailboxes live in a sibling directory of the bucket store, not inside it. Two
  reasons, both load-bearing: `Sync.Backend.Sweeper` treats every directory under
  the bucket root as a bucket to run snapshot retention over, and bucket names
  are client-controlled — a client could PUT to a bucket named like whatever
  reserved subdirectory we picked, and reach these files through the wrong routes.
  """

  alias Sync.Backend.Store

  # The six slots of `roam_pairing::mailbox::Slot`, and the only names accepted.
  # An allowlist rather than a charset check: this is a pairing mailbox, not a
  # general-purpose object store, and it should be impossible to use it as one.
  @slots ~w(msg1 msg2 confirm-joiner confirm-host request accept)

  # Rendezvous and session ids are always 32 bytes of base64url. Requiring the
  # exact length (rather than Store.valid_id?'s "<= 128 chars") bounds directory
  # names and rejects anything that could not have come from the protocol.
  @id_length 43

  # One accept carries the host's transitive roster and key log, which grow with
  # the number of devices and role changes. Generous enough not to cap a real
  # vault, bounded so an unauthenticated writer cannot fill a disk with one PUT.
  @max_body_bytes 4_000_000

  # Enough for a joiner to retry, and for a few squatters, without letting anyone
  # create directories under one rendezvous without limit.
  @max_sessions 64

  # A pairing window is five minutes; anything untouched for this long is dead.
  @default_ttl_ms 30 * 60 * 1000

  @doc "The slot names this mailbox accepts."
  def slots, do: @slots

  @doc "Maximum body size for one slot, in bytes."
  def max_body_bytes, do: @max_body_bytes

  @doc "Maximum concurrent sessions under one rendezvous."
  def max_sessions, do: @max_sessions

  @doc """
  Absolute mailbox root — a *sibling* of the bucket store, never inside it. See
  the module note on why that separation matters.
  """
  def data_root do
    Application.get_env(:sync, :mailbox_data_root) || Store.data_root() <> "-rendezvous"
  end

  @doc "Whether `id` could be a rendezvous or session id."
  @spec valid_id?(String.t()) :: boolean()
  def valid_id?(id),
    do: is_binary(id) and byte_size(id) == @id_length and Store.valid_id?(id)

  @doc "Whether `slot` is one of the six handshake slots."
  @spec valid_slot?(String.t()) :: boolean()
  def valid_slot?(slot), do: slot in @slots

  @doc """
  Write a slot.

  `:created` on the first write, `:exists` if it was already taken (the stored
  body is left alone), `:too_many_sessions` if this would open a new session past
  the cap.
  """
  @spec put(String.t(), String.t(), String.t(), binary()) ::
          :created | :exists | :too_many_sessions
  def put(rendezvous, session, slot, bytes) do
    session_dir = Path.join([data_root(), rendezvous, session])
    path = Path.join(session_dir, slot)

    cond do
      File.exists?(path) ->
        :exists

      not File.dir?(session_dir) and session_count(rendezvous) >= @max_sessions ->
        :too_many_sessions

      true ->
        File.mkdir_p!(session_dir)
        # Write to a temporary name first, so a reader never observes a
        # half-written slot.
        tmp = path <> ".tmp." <> Integer.to_string(System.unique_integer([:positive]))
        File.write!(tmp, bytes)

        # THIS is what enforces write-once, not the `File.exists?` above — that
        # one is only a fast path that avoids a pointless write for a repeat
        # writer, and two concurrent requests can both sail past it. `File.ln`
        # fails if the target exists, atomically, which is exactly the semantics
        # a write-once slot needs; `File.rename` would silently clobber.
        #
        # Mutation-checked, and worth recording because the result was not what
        # it looks like: removing EITHER guard alone breaks nothing, because
        # sequentially they cover for each other. Only removing both fails
        # `a taken slot refuses the second write`, and only a concurrent test
        # (`two simultaneous writers`) distinguishes the link from the rename.
        case File.ln(tmp, path) do
          :ok ->
            File.rm(tmp)
            :created

          {:error, _already_there_or_unsupported} ->
            File.rm(tmp)
            if File.exists?(path), do: :exists, else: slow_path_write(tmp, path, bytes)
        end
    end
  end

  # Filesystems without hard links (rare, but not worth crashing on) fall back to
  # rename. The race window reopens there; on such a filesystem two simultaneous
  # writers to one slot can let the second win. Stated rather than pretended
  # away — every deployment target has links.
  defp slow_path_write(tmp, path, bytes) do
    File.write!(tmp, bytes)
    File.rename!(tmp, path)
    :created
  end

  @doc "The bytes in a slot, or nil."
  @spec get(String.t(), String.t(), String.t()) :: binary() | nil
  def get(rendezvous, session, slot) do
    case File.read(Path.join([data_root(), rendezvous, session, slot])) do
      {:ok, bytes} -> bytes
      {:error, _} -> nil
    end
  end

  @doc "Session ids present under `rendezvous`."
  @spec sessions(String.t()) :: [String.t()]
  def sessions(rendezvous) do
    dir = Path.join(data_root(), rendezvous)

    case File.ls(dir) do
      {:ok, names} -> names |> Enum.filter(&valid_id?/1) |> Enum.sort()
      {:error, _} -> []
    end
  end

  defp session_count(rendezvous), do: length(sessions(rendezvous))

  @doc """
  Delete every rendezvous untouched for longer than the TTL.

  Pairing is an interactive, user-present action measured in minutes, so there is
  nothing here worth keeping. Returns the ids removed.
  """
  @spec sweep(keyword()) :: [String.t()]
  def sweep(opts \\ []) do
    root = Keyword.get(opts, :data_root, data_root())
    ttl = Keyword.get(opts, :ttl_ms, Application.get_env(:sync, :mailbox_ttl_ms, @default_ttl_ms))
    now = Keyword.get(opts, :now_ms, System.system_time(:millisecond))

    case File.ls(root) do
      {:ok, names} ->
        for name <- names,
            path = Path.join(root, name),
            File.dir?(path),
            stale?(path, now, ttl) do
          File.rm_rf(path)
          name
        end

      {:error, _} ->
        []
    end
  end

  # A rendezvous directory's mtime moves every time a session is created under
  # it, so this is "no new session for a while". Slots written inside an existing
  # session do not touch it — which is fine, since the TTL is six times the
  # longest pairing window.
  defp stale?(path, now, ttl) do
    case File.stat(path, time: :posix) do
      {:ok, %{mtime: mtime}} -> now - mtime * 1000 >= ttl
      _ -> false
    end
  end
end

File: crates/roam-pairing/src/mailbox.rs
//! The mailbox: six write-once slots per session, and nothing else.
//!
//! Pairing needs a two-way channel between devices that may never be able to
//! reach each other directly — a browser tab and a phone, say. The relay already
//! stands between them, so it can carry the handshake, but only if it is given a
//! shape that keeps it as ignorant as it is for everything else it stores.
//!
//! That shape is this: a **rendezvous** (named by 32 unguessable bytes the host
//! mints) holds **sessions** (named by 32 bytes a joiner mints), and each session
//! holds six named slots. A slot is written once. There is no delete, no list of
//! slot contents, no way to ask the relay anything about a body it holds.
//!
//! # What write-once actually buys, which is not what it looks like
//!
//! The obvious answer is that it stops a relay rewriting a handshake message to
//! split the two sides' views of the transcript. That is *not* what is doing the
//! work: the confirmations in [`roam_pake`] are MACs over both SPAKE2 messages,
//! so a rewritten message fails the confirmation whether or not the write was
//! allowed. Tampering is caught by the MACs, and it is caught with a dishonest
//! relay too — which write-once could never help with, since a relay that is
//! willing to rewrite is willing to ignore its own rule.
//!
//! What write-once actually buys is that **squatting is free rather than
//! fatal**. Anyone who knows the rendezvous can open a session and write the slot
//! the *host* was going to write, plus a garbage confirmation. A host that
//! shrugged at the taken slot and carried on would then verify a confirmation
//! against a transcript it never wrote — which fails, and spends one of three
//! attempts. Three of those retire the code without the squatter guessing a
//! single digit. Refusing to continue past a taken slot is what makes that cost
//! nothing. See `squatting_a_slot_costs_the_host_no_attempts`, which fails if
//! either half of that (the relay's refusal or the host's) is removed.
//!
//! It also means a session is *consumed*: a peer that grabs a session id and
//! writes rubbish into it has spoiled that session and no other. The joiner mints
//! a fresh one and tries again, which is why sessions exist at all rather than
//! the six slots living directly under the rendezvous.
//!
//! # What the relay learns
//!
//! A rendezvous id, some session ids, and six opaque bodies per session. Not the
//! bucket, not the vault, not either device's key: the SPAKE2 identities are the
//! session and rendezvous ids the relay already routes on, and everything after
//! the handshake is sealed under a key derived from a code the relay never sees.
//! See [`crate::handshake`] for why that particular choice of identity strings.

use crate::MaybeSendSync;
use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Mutex;

/// The six messages one handshake exchanges, and the only slot names a relay
/// accepts.
///
/// An enum rather than a free string so the wire names live in one place and a
/// caller cannot invent a seventh — the relay enforces the same allowlist, and
/// two allowlists that can drift are one allowlist too many.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Slot {
    /// Joiner → host: the SPAKE2 first message.
    Msg1,
    /// Host → joiner: the SPAKE2 second message.
    Msg2,
    /// Joiner → host: the joiner's key confirmation. It proves first.
    ConfirmJoiner,
    /// Host → joiner: the host's key confirmation, withheld until the joiner's
    /// verifies — it is the only oracle a guesser would have.
    ConfirmHost,
    /// Joiner → host, sealed: which device it is asking to enrol.
    Request,
    /// Host → joiner, sealed: the vault key, rosters, key log and founder pin.
    Accept,
}

impl Slot {
    pub fn as_str(self) -> &'static str {
        match self {
            Slot::Msg1 => "msg1",
            Slot::Msg2 => "msg2",
            Slot::ConfirmJoiner => "confirm-joiner",
            Slot::ConfirmHost => "confirm-host",
            Slot::Request => "request",
            Slot::Accept => "accept",
        }
    }

    /// Every slot, so a relay implementation (or a test double) can build its
    /// allowlist from the protocol rather than from a copied list.
    pub const ALL: [Slot; 6] = [
        Slot::Msg1,
        Slot::Msg2,
        Slot::ConfirmJoiner,
        Slot::ConfirmHost,
        Slot::Request,
        Slot::Accept,
    ];

    pub fn parse(name: &str) -> Option<Slot> {
        Slot::ALL.into_iter().find(|slot| slot.as_str() == name)
    }
}

/// What a write met when it landed.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SlotOutcome {
    Written,
    /// Somebody wrote here first. The caller's message did NOT land, and the
    /// slot still holds whatever it held — this is never "close enough".
    AlreadyTaken,
}

/// A rendezvous, as seen by one device. The rendezvous id is fixed when the
/// implementation is constructed, so it cannot vary per call.
#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
pub trait Mailbox: MaybeSendSync {
    /// Write a slot. Must not overwrite: a slot that already holds bytes returns
    /// [`SlotOutcome::AlreadyTaken`] and the stored body is untouched.
    async fn put(&self, session: &str, slot: Slot, body: Vec<u8>) -> anyhow::Result<SlotOutcome>;

    /// Read a slot, or `None` if it has not been written yet. Absence is the
    /// normal case while polling, not an error.
    async fn get(&self, session: &str, slot: Slot) -> anyhow::Result<Option<Vec<u8>>>;

    /// Session ids present under this rendezvous. The host polls this to notice
    /// a joiner; the joiner never needs it.
    async fn sessions(&self) -> anyhow::Result<Vec<String>>;
}

/// An in-process mailbox, for tests and for a same-device flow.
///
/// Shared by cloning: every clone addresses the same slots, which is what lets a
/// test hand one end to a host and the other to a joiner without a relay.
type Slots = std::sync::Arc<Mutex<HashMap<(String, Slot), Vec<u8>>>>;

#[derive(Clone, Default)]
pub struct MemoryMailbox {
    slots: Slots,
}

impl MemoryMailbox {
    pub fn new() -> Self {
        Self::default()
    }

    /// Overwrite a slot regardless of whether it is taken — the one thing the
    /// trait promises never to do, exposed so a test can play a relay that
    /// breaks its promise. Not available to protocol code.
    #[doc(hidden)]
    pub fn force(&self, session: &str, slot: Slot, body: Vec<u8>) {
        self.slots
            .lock()
            .expect("mailbox mutex poisoned")
            .insert((session.to_string(), slot), body);
    }
}

#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
impl Mailbox for MemoryMailbox {
    async fn put(&self, session: &str, slot: Slot, body: Vec<u8>) -> anyhow::Result<SlotOutcome> {
        let mut slots = self.slots.lock().expect("mailbox mutex poisoned");
        let key = (session.to_string(), slot);
        if slots.contains_key(&key) {
            return Ok(SlotOutcome::AlreadyTaken);
        }
        slots.insert(key, body);
        Ok(SlotOutcome::Written)
    }

    async fn get(&self, session: &str, slot: Slot) -> anyhow::Result<Option<Vec<u8>>> {
        Ok(self
            .slots
            .lock()
            .expect("mailbox mutex poisoned")
            .get(&(session.to_string(), slot))
            .cloned())
    }

    async fn sessions(&self) -> anyhow::Result<Vec<String>> {
        let slots = self.slots.lock().expect("mailbox mutex poisoned");
        let mut names: Vec<String> = slots.keys().map(|(session, _)| session.clone()).collect();
        names.sort();
        names.dedup();
        Ok(names)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn every_slot_round_trips_through_its_wire_name() {
        // The relay parses these names from a URL path and this enum writes
        // them. A name that serialises to something `parse` does not accept is a
        // slot the two halves disagree about, which shows up as a hung
        // handshake rather than an error.
        for slot in Slot::ALL {
            assert_eq!(Slot::parse(slot.as_str()), Some(slot));
        }
    }

    #[test]
    fn slot_names_are_all_distinct() {
        // Two slots sharing a name would silently alias, so the host would read
        // its own message back as the joiner's.
        let mut names: Vec<&str> = Slot::ALL.iter().map(|slot| slot.as_str()).collect();
        names.sort_unstable();
        let count = names.len();
        names.dedup();
        assert_eq!(names.len(), count, "two slots share a wire name");
    }

    #[tokio::test]
    async fn a_written_slot_refuses_a_second_write_and_keeps_the_first() {
        // Write-once is what stops a relay rewriting a handshake message after
        // the other side has already read and MAC'd it. "Refused" is not enough
        // on its own — the original bytes must still be what a reader gets.
        let mailbox = MemoryMailbox::new();
        assert_eq!(
            mailbox
                .put("s", Slot::Msg1, b"first".to_vec())
                .await
                .unwrap(),
            SlotOutcome::Written
        );
        assert_eq!(
            mailbox
                .put("s", Slot::Msg1, b"second".to_vec())
                .await
                .unwrap(),
            SlotOutcome::AlreadyTaken
        );
        assert_eq!(
            mailbox.get("s", Slot::Msg1).await.unwrap(),
            Some(b"first".to_vec())
        );
    }

    #[tokio::test]
    async fn slots_and_sessions_do_not_bleed_into_each_other() {
        let mailbox = MemoryMailbox::new();
        mailbox.put("a", Slot::Msg1, b"a1".to_vec()).await.unwrap();
        mailbox.put("a", Slot::Msg2, b"a2".to_vec()).await.unwrap();
        mailbox.put("b", Slot::Msg1, b"b1".to_vec()).await.unwrap();

        assert_eq!(
            mailbox.get("a", Slot::Msg1).await.unwrap(),
            Some(b"a1".to_vec())
        );
        assert_eq!(
            mailbox.get("b", Slot::Msg1).await.unwrap(),
            Some(b"b1".to_vec())
        );
        assert_eq!(mailbox.get("b", Slot::Msg2).await.unwrap(), None);
        assert_eq!(
            mailbox.sessions().await.unwrap(),
            vec!["a".to_string(), "b".to_string()]
        );
    }
}
